src/ZSingleWireSerial.cpp

```diff
@@ -0,0 +1,287 @@
+        while(SERCOM0->USART.SYNCBUSY.bit.ENABLE);
+
+        SERCOM0->USART.CTRLB.bit.TXEN = 1;
+        while(SERCOM0->USART.SYNCBUSY.bit.CTRLB);
+        status |= TX_CONFIGURED;
+    }
+    else if (status & TX_CONFIGURED && !enable)
+    {
+        SERCOM0->USART.CTRLB.bit.TXEN = 0;
+        while(SERCOM0->USART.SYNCBUSY.bit.CTRLB);
+        //
+        SERCOM0->USART.CTRLA.bit.ENABLE = 0;
+        while(SERCOM0->USART.SYNCBUSY.bit.ENABLE);
+
+        gpio_set_pin_function(p.name, GPIO_PIN_FUNCTION_OFF);
+        status &= ~TX_CONFIGURED;
+    }
+
+    return DEVICE_OK;
+}
+
+int ZSingleWireSerial::configureRx(int enable)
+{
+    if (enable && !(status & RX_CONFIGURED))
+    {
+        gpio_set_pin_function(p.name, PINMUX_PA04D_SERCOM0_PAD0);
+
+        SERCOM0->USART.CTRLA.bit.ENABLE = 0;
+        while(SERCOM0->USART.SYNCBUSY.bit.ENABLE);
+
+        SERCOM0->USART.CTRLA.bit.SAMPR = 0;
+        SERCOM0->USART.CTRLA.bit.RXPO = 0; // PAD 0
+        SERCOM0->USART.CTRLB.bit.CHSIZE = 0; // 8 BIT
+
+        SERCOM0->USART.CTRLA.bit.ENABLE = 1;
+        while(SERCOM0->USART.SYNCBUSY.bit.ENABLE);
+
+        SERCOM0->USART.CTRLB.bit.RXEN = 1;
+        while(SERCOM0->USART.SYNCBUSY.bit.CTRLB);
+
+        status |= RX_CONFIGURED;
+    }
+    else if (status & RX_CONFIGURED)
+    {
+        SERCOM0->USART.CTRLB.bit.RXEN = 0;
+        while(SERCOM0->USART.SYNCBUSY.bit.CTRLB);
+        //
+        SERCOM0->USART.CTRLA.bit.ENABLE = 0;
+        while(SERCOM0->USART.SYNCBUSY.bit.ENABLE);
+
+        gpio_set_pin_function(p.name, GPIO_PIN_FUNCTION_OFF);
+        status &= ~RX_CONFIGURED;
+    }
+
+    return DEVICE_OK;
+}
+
+int ZSingleWireSerial::setMode(SingleWireMode sw)
+{
+    if (sw == SingleWireRx)
+    {
+        configureTx(0);
+        configureRx(1);
+    }
+    else if (sw == SingleWireTx)
+    {
+        configureRx(0);
+        configureTx(1);
+    }
+    else
+    {
+        configureTx(0);
+        configureRx(0);
+    }
+
+    return DEVICE_OK;
+}
+
+int ZSingleWireSerial::send(uint8_t* data, int len)
+{
+    if (!(status & TX_CONFIGURED))
+        setMode(SingleWireTx);
+
+    for (int i = 0; i < len; i++)
+        putc(data[i]);
+
+    return DEVICE_OK;
+}
+
+int ZSingleWireSerial::receive(uint8_t* data, int len)
+{
+    if (!(status & RX_CONFIGURED))
+        setMode(SingleWireRx);
+
+    for (int i = 0; i < len; i++)
+        data[i] = getc();
+
+    return DEVICE_OK;
+}
+
+int ZSingleWireSerial::sendDMA(uint8_t* data, int len)
+{
+    if (!(status & RX_CONFIGURED))
+        setMode(SingleWireTx);
+
+    usart_dma->configure(sercom_trigger_src(0, true), BeatByte, NULL, (volatile void*)&SERCOM0->USART.DATA.reg);
+    usart_dma->transfer((const void*)data, NULL, len);
+    return DEVICE_OK;
+}
+
+int ZSingleWireSerial::receiveDMA(uint8_t* data, int len)
+{
+    if (!(status & RX_CONFIGURED))
+        setMode(SingleWireRx);
+
+    usart_dma->configure(sercom_trigger_src(0, false), BeatByte, (volatile void*)SERCOM0->USART.DATA.reg, NULL);
```

Comment:
And a DMA instance in use.

Reply:
Maybe better to use two DMA channels? I think we have plenty of them. Then you wouldn't need to re-configure them, which might save some cycles (esp. on SAMD21)?

Reply:
Absolutely, just trying to get it running at the moment :smile:

Any idea on why I get 11 completion events for a 4 byte payload? I'm bringing up another serial debug line to work out why this is happening.

Reply:
I was getting only one for SPI (I think)...